tsl/src/compression/compression.c

```diff
@@ -157,6 +157,51 @@ static void row_compressor_finish(RowCompressor *row_compressor);
  ** compress_chunk **
  ********************/
 
+static void
+capture_pgclass_stats(Relation rel, int *out_pages, int *out_visible, float *out_tuples)
+{
+	Relation pg_class = table_open(RelationRelationId, RowExclusiveLock);
+	HeapTuple tuple = SearchSysCacheCopy1(RELOID, ObjectIdGetDatum(RelationGetRelid(rel)));
+	Form_pg_class classform;
+
+	if (!HeapTupleIsValid(tuple))
+		elog(ERROR, "could not find tuple for relation %u", RelationGetRelid(rel));
+
+	classform = (Form_pg_class) GETSTRUCT(tuple);
+
+	*out_pages = classform->relpages;
+	*out_visible = classform->relallvisible;
+	*out_tuples = classform->reltuples;
+
+	heap_freetuple(tuple);
+	table_close(pg_class, RowExclusiveLock);
+}
+
+static void
+restore_pgclass_stats(Relation rel, int pages, int visible, float tuples)
+{
+	Relation pg_class;
+	HeapTuple tuple;
+	Form_pg_class classform;
+
+	pg_class = table_open(RelationRelationId, RowExclusiveLock);
+	tuple = SearchSysCacheCopy1(RELOID, ObjectIdGetDatum(RelationGetRelid(rel)));
+	if (!HeapTupleIsValid(tuple))
+		elog(ERROR, "could not find tuple for relation %u", RelationGetRelid(rel));
+	classform = (Form_pg_class) GETSTRUCT(tuple);
+
+	classform->relpages = pages;
+	classform->relallvisible = visible;
+	classform->reltuples = tuples;
+
+	CatalogTupleUpdate(pg_class, &tuple->t_self, tuple);
+
+	heap_freetuple(tuple);
+	table_close(pg_class, RowExclusiveLock);
+
+	CommandCounterIncrement();
```

Comment:
Often it is better to do this outside the functions since the caller might want to make multiple changes visible in one single CommandCounterIncrement

Reply:
fixed